src/Compilers/CSharp/Portable/Syntax/NonNullDirectiveMap.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Syntax
+{
+    internal sealed class NonNullDirectiveMap
+    {
+        private static readonly NonNullDirectiveMap Empty = new NonNullDirectiveMap(ImmutableArray<(int Position, bool State)>.Empty);
+
+        private readonly ImmutableArray<(int Position, bool State)> _directives;
+
+        internal static NonNullDirectiveMap Create(SyntaxTree tree)
+        {
+            var directives = GetDirectives(tree);
+            return directives.IsEmpty ? Empty : new NonNullDirectiveMap(directives);
+        }
+
+        private NonNullDirectiveMap(ImmutableArray<(int Position, bool State)> directives)
+        {
+#if DEBUG
+            for (int i = 1; i < directives.Length; i++)
+            {
+                Debug.Assert(directives[i - 1].Position < directives[i].Position);
+            }
+#endif
+            _directives = directives;
+        }
+
+        /// <summary>
+        /// Returns true if the `#nonnull` directive preceding the position is
+        /// `restore`, false for if `disable`, and null if no preceding directive.
+        /// </summary>
+        internal bool? GetDirectiveState(int position)
+        {
+            int index = _directives.BinarySearch((position, false), PositionComparer.Instance);
+            if (index < 0)
+            {
+                index = ~index - 1;
```

Comment:
>index = ~index - 1; [](start = 16, length = 19)

I didn't understand the purpose of this.

Reply:
If there is not an exact match, `BinarySearch` returns the complement of the nearest index.

---
In reply to: [225352180](https://github.com/dotnet/roslyn/pull/30499#discussion_r225352180) [](ancestors = 225352180)